src/main/java/uk/gov/ch/developer/docs/controller/developer/UserCallbackController.java

```diff
@@ -41,22 +52,90 @@ public String getCallback(@RequestParam("state") String state,
+        LOGGER.debug("Getting User Profile....");
+
+        OAuthToken oauthToken = oauthTokenResponse.block();
+
+        return requestUserProfile(signInData, webClient, oauthToken);
+    }
+
+    private UserProfileResponse requestUserProfile(Map<String, Object> signInData, WebClient webClient,
+            OAuthToken tokenResponse) {
+        setAccessToken(signInData, tokenResponse);
+        return webClient.get().uri(URI.create(identityProvider.getProfileUrl()))
+                .headers(h -> h.setBearerAuth(tokenResponse.getToken()))
+                .accept(MediaType.APPLICATION_JSON).exchange()
+                .flatMap(response -> response.bodyToMono(UserProfileResponse.class))
+                .block();
+    }
+
+    private Mono<OAuthToken> getOAuthToken(String code, WebClient webClient) {
+        LOGGER.debug("Getting OAuth Token....");
+        return webClient.post().uri(URI.create(identityProvider.getTokenUrl()))
+                .contentType(MediaType.APPLICATION_FORM_URLENCODED)
+                .accept(MediaType.APPLICATION_JSON).body(BodyInserters
+                        .fromPublisher(Mono.just(getPostRequestBody(code)), String.class))
+                .retrieve()
+                .bodyToMono(OAuthToken.class);
+    }
+
+    private void setUserProfile(Map<String, Object> signInData, UserProfileResponse userProfileResponse) {
+        LOGGER.debug("Setting User Profile...");
+        Map<String, Object> userProfileData = new HashMap<>();
+
+        userProfileData.put(SessionKeys.EMAIL.getKey(), userProfileResponse.getEmail());
+        userProfileData.put(SessionKeys.USER_ID.getKey(), userProfileResponse.getId());
+        userProfileData.put(SessionKeys.LOCALE.getKey(), userProfileResponse.getLocale());
+        userProfileData.put(SessionKeys.SCOPE.getKey(), userProfileResponse.getScope());
+        userProfileData.put(SessionKeys.FORENAME.getKey(), userProfileResponse.getForename());
+        userProfileData.put(SessionKeys.SURNAME.getKey(), userProfileResponse.getSurname());
+        userProfileData.put(SessionKeys.PERMISSIONS.getKey(), userProfileResponse.getPermissions());
+
+        LOGGER.debugContext("User Profile Data: ", "setUserProfile() Method", userProfileData);
+
+        signInData.put(SessionKeys.USER_PROFILE.getKey(), userProfileData);
+    }
+
+    private void setAccessToken(Map<String, Object> signInData, OAuthToken token) {
+        LOGGER.debug("Setting access token...");
+        Map<String, Object> accessTokenData = new HashMap<>();
+
+        accessTokenData.put(SessionKeys.ACCESS_TOKEN.getKey(), token.getToken());
+        accessTokenData.put(SessionKeys.EXPIRES_IN.getKey(), token.getExpiresIn());
+        accessTokenData.put(SessionKeys.REFRESH_TOKEN.getKey(), token.getRefreshToken());
+        accessTokenData.put(SessionKeys.TOKEN_TYPE.getKey(), token.getTokenType());
+
+        signInData.put(SessionKeys.ACCESS_TOKEN.getKey(), accessTokenData);
+    }
+
```

Comment:
Most of the `getPostRequestBody` looks like it should be delegated to the identity provider or OAuth instance.